File: packages/tui/src/data/git/blame.ts
import { Effect, Option } from "effect";
import type { BlameTarget } from "#tui/types.ts";
import {
	type CommitDiffSelection,
	EMPTY_TREE_HASH,
	GitCommandError,
	runGitEffectAsync,
} from "#data/git/core.ts";

const FIELD_SEPARATOR = "\u001f";
const UNCOMMITTED_BLAME_HASH = "0000000000000000000000000000000000000000";

export interface ParsedBlameHeader {
	readonly commitHash: string;
	readonly author: string;
	readonly date: string;
	readonly summary: string;
}

export interface ParsedCommitShow {
	readonly commitHash: string;
	readonly shortHash: string;
	readonly parentHashes: ReadonlyArray<string>;
	readonly date: string;
	readonly author: string;
	readonly subject: string;
	readonly description: string;
}

export interface BlameCommitDetails {
	readonly target: BlameTarget;
	readonly commitHash: string;
	readonly shortHash: string;
	readonly author: string;
	readonly date: string;
	readonly subject: string;
	readonly description: string;
	readonly isUncommitted: boolean;
	readonly compareSelection: Option.Option<CommitDiffSelection>;
}

export function isUncommittedBlameHash(hash: string): boolean {
	return hash.trim() === UNCOMMITTED_BLAME_HASH;
}

function formatUnixDate(rawSeconds: string): string {
	const seconds = Number(rawSeconds);
	if (!Number.isFinite(seconds) || seconds <= 0) {
		return "";
	}
	const date = new Date(seconds * 1000);
	return Number.isNaN(date.getTime()) ? "" : date.toISOString().slice(0, 10);
}

export function parseBlamePorcelainHeader(
	rawOutput: string,
): Option.Option<ParsedBlameHeader> {
	const lines = rawOutput.split("\n");
	const firstLine = lines[0]?.trim() ?? "";
	const firstLineFields = firstLine.split(/\s+/);
	const commitHash = firstLineFields[0]?.trim() ?? "";
	if (commitHash.length !== 40) {
		return Option.none();
	}

	let author = "";
	let date = "";
	let summary = "";

	for (const line of lines.slice(1)) {
		if (line.startsWith("\t")) {
			break;
		}
		if (line.startsWith("author ")) {
			author = line.slice("author ".length).trim();
			continue;
		}
		if (line.startsWith("author-time ")) {
			date = formatUnixDate(line.slice("author-time ".length).trim());
			continue;
		}
		if (line.startsWith("summary ")) {
			summary = line.slice("summary ".length).trim();
		}
	}

	return Option.some({
		commitHash,
		author,
		date,
		summary,
	});
}

export function parseCommitShowOutput(
	rawOutput: string,
): Option.Option<ParsedCommitShow> {
	const [
		commitHash = "",
		shortHash = "",
		parentsRaw = "",
		date = "",
		author = "",
		subject = "",
		description = "",
	] = rawOutput.split(FIELD_SEPARATOR);

	if (commitHash.length === 0 || shortHash.length === 0) {
		return Option.none();
	}

	return Option.some({
		commitHash,
		shortHash,
		parentHashes: parentsRaw
			.split(" ")
			.map((hash) => hash.trim())
			.filter((hash) => hash.length > 0),
		date: date.trim(),
		author: author.trim(),
		subject: subject.trim(),
		description: description.trim(),
	});
}

function buildParseError(
	args: ReadonlyArray<string>,
	stdout: string,
	stderr: string,
	fallbackMessage: string,
): GitCommandError {
	return new GitCommandError({
		args: [...args],
		stdout,
		stderr,
		fallbackMessage,
	});
}

export function loadBlameCommitDetails(
	target: BlameTarget,
): Effect.Effect<BlameCommitDetails, GitCommandError> {
	const blameArgs = [
		"blame",
		"--porcelain",
		"-L",
		`${target.lineNumber},${target.lineNumber}`,
		"--",
		target.filePath,
	];

	return Effect.gen(function* () {
		const blameResult = yield* runGitEffectAsync(
			blameArgs,
			`Unable to load blame for ${target.filePath}:${target.lineNumber}.`,
		);
		const parsedHeader = parseBlamePorcelainHeader(blameResult.stdout);

		if (Option.isNone(parsedHeader)) {
			return yield* Effect.fail(
				buildParseError(
					blameArgs,
					blameResult.stdout,
					blameResult.stderr,
					`Unable to parse blame output for ${target.filePath}:${target.lineNumber}.`,
				),
			);
		}

		const header = parsedHeader.value;
		if (isUncommittedBlameHash(header.commitHash)) {
			return {
				target,
				commitHash: header.commitHash,
				shortHash: "working-tree",
				author: header.author.length > 0 ? header.author : "Uncommitted",
				date: header.date,
				subject:
					header.summary.length > 0
						? header.summary
						: "Uncommitted line changes",
				description:
					"This line has uncommitted changes. Commit comparison is unavailable.",
				isUncommitted: true,
				compareSelection: Option.none(),
			} satisfies BlameCommitDetails;
		}

		const showArgs = [
			"show",
			"-s",
			"--date=short",
			`--format=%H${FIELD_SEPARATOR}%h${FIELD_SEPARATOR}%P${FIELD_SEPARATOR}%ad${FIELD_SEPARATOR}%an${FIELD_SEPARATOR}%s${FIELD_SEPARATOR}%b`,
			header.commitHash,
		];
		const showResult = yield* runGitEffectAsync(
			showArgs,
			`Unable to load commit metadata for ${header.commitHash}.`,
		);
		const parsedShow = parseCommitShowOutput(showResult.stdout);
		if (Option.isNone(parsedShow)) {
			return yield* Effect.fail(
				buildParseError(
					showArgs,
					showResult.stdout,
					showResult.stderr,
					`Unable to parse commit metadata for ${header.commitHash}.`,
				),
			);
		}

		const commit = parsedShow.value;
		const compareSelection: CommitDiffSelection = {
			commitHash: commit.commitHash,
			baseRef: commit.parentHashes[0] ?? EMPTY_TREE_HASH,
			shortHash: commit.shortHash,
			subject: commit.subject.length > 0 ? commit.subject : header.summary,
		};

		return {
			target,
			commitHash: commit.commitHash,
			shortHash: commit.shortHash,
			author: commit.author.length > 0 ? commit.author : header.author,
			date: commit.date.length > 0 ? commit.date : header.date,
			subject: commit.subject.length > 0 ? commit.subject : header.summary,
			description: commit.description,
			isUncommitted: false,
			compareSelection: Option.some(compareSelection),
		} satisfies BlameCommitDetails;
	});
}

File: src/git/commit/blame.rs
use std::path::Path;

use color_eyre::eyre::eyre;

use super::super::{
    BlameCommitDetails, BlameTarget, CommitCompareSelection, EMPTY_TREE_HASH,
    command::git_output,
    parse::{is_uncommitted_blame_hash, parse_blame_porcelain_header, parse_commit_show_output},
};

pub async fn load_blame_commit_details(
    repo_root: &Path,
    target: &BlameTarget,
) -> color_eyre::Result<BlameCommitDetails> {
    let blame_output = git_output(
        repo_root,
        &[
            "blame",
            "--porcelain",
            "-L",
            &format!("{0},{0}", target.line_number),
            "--",
            target.file_path.as_str(),
        ],
    )
    .await?;

    let header = parse_blame_porcelain_header(&blame_output).ok_or_else(|| {
        eyre!(
            "unable to parse blame output for {}:{}",
            target.file_path,
            target.line_number
        )
    })?;

    if is_uncommitted_blame_hash(&header.commit_hash) {
        return Ok(BlameCommitDetails {
            target: target.clone(),
            commit_hash: header.commit_hash,
            short_hash: "working-tree".to_string(),
            author: if header.author.is_empty() {
                "Uncommitted".to_string()
            } else {
                header.author
            },
            date: header.date,
            subject: if header.summary.is_empty() {
                "Uncommitted line changes".to_string()
            } else {
                header.summary
            },
            description: "This line has uncommitted changes. Commit comparison is unavailable."
                .to_string(),
            is_uncommitted: true,
            compare_selection: None,
        });
    }

    let show_output = git_output(
        repo_root,
        &[
            "show",
            "-s",
            "--date=short",
            "--format=%H%x1f%h%x1f%P%x1f%ad%x1f%an%x1f%s%x1f%b",
            header.commit_hash.as_str(),
        ],
    )
    .await?;

    let commit = parse_commit_show_output(&show_output)
        .ok_or_else(|| eyre!("unable to parse commit metadata for {}", header.commit_hash))?;
    let subject = if commit.subject.is_empty() {
        header.summary
    } else {
        commit.subject
    };
    let description = if commit.description.trim().is_empty() {
        "No commit description.".to_string()
    } else {
        commit.description
    };
    let compare_base = commit
        .parent_hashes
        .first()
        .cloned()
        .unwrap_or_else(|| EMPTY_TREE_HASH.to_string());
    let commit_hash = commit.commit_hash;
    let short_hash = commit.short_hash;

    Ok(BlameCommitDetails {
        target: target.clone(),
        commit_hash: commit_hash.clone(),
        short_hash: short_hash.clone(),
        author: if commit.author.is_empty() {
            header.author
        } else {
            commit.author
        },
        date: if commit.date.is_empty() {
            header.date
        } else {
            commit.date
        },
        description,
        is_uncommitted: false,
        compare_selection: Some(CommitCompareSelection {
            base_ref: compare_base,
            commit_hash: commit_hash.clone(),
            short_hash: short_hash.clone(),
            subject: subject.clone(),
        }),
        subject,
    })
}
